src/patches.js

```diff
@@ -6,50 +6,92 @@ export function generatePatches(state, basePath, patches, inversePatches) {
         : generateObjectPatches(state, basePath, patches, inversePatches)
 }
 
-function generateArrayPatches(state, basePath, patches, inversePatches) {
-    const {base, copy, assigned} = state
-    const minLength = Math.min(base.length, copy.length)
+function _generateArrayPatches(
+    base,
+    copy,
+    assigned,
+    basePath,
+    patches,
+    inversePatches
+) {
+    if (copy.length < base.length) {
+        return _generateArrayPatches(
+            copy,
+            base,
+            assigned,
+            basePath,
+            inversePatches,
+            patches
+        )
+    }
 
-    // Look for replaced indices.
-    for (let i = 0; i < minLength; i++) {
-        if (assigned[i] && base[i] !== copy[i]) {
-            const path = basePath.concat(i)
-            patches.push({op: "replace", path, value: copy[i]})
-            inversePatches.push({op: "replace", path, value: base[i]})
-        }
+    const delta = copy.length - base.length
+
+    let start = 0
+    while (base[start] === copy[start] && start < base.length) {
+        ++start
+    }
+
+    let baseEnd = base.length
+    while (baseEnd > start && base[baseEnd - 1] === copy[baseEnd + delta - 1]) {
+        --baseEnd
     }
 
-    // Did the array expand?
-    if (minLength < copy.length) {
-        for (let i = minLength; i < copy.length; i++) {
+    const replaceCount = baseEnd - start
+    for (let i = 0; i < replaceCount; ++i) {
+        const index = start + i
+        const path = basePath.concat([index])
+        if (assigned[index] && copy[index] !== base[index]) {
             patches.push({
-                op: "add",
-                path: basePath.concat(i),
-                value: copy[i]
+                op: "replace",
+                path,
+                value: copy[index]
+            })
+            inversePatches.push({
+                op: "replace",
+                path,
+                value: base[index]
             })
         }
-        inversePatches.push({
-            op: "replace",
-            path: basePath.concat("length"),
-            value: base.length
-        })
     }
+    start += replaceCount
 
-    // ...or did it shrink?
-    else if (minLength < base.length) {
+    const useRemove = start != base.length
+    for (let j = 0; j < delta; ++j) {
+        const path = basePath.concat([start + j])
         patches.push({
-            op: "replace",
-            path: basePath.concat("length"),
-            value: copy.length
+            op: "add",
+            path,
+            value: copy[start + j]
         })
-        for (let i = minLength; i < base.length; i++) {
-            inversePatches.push({
-                op: "add",
-                path: basePath.concat(i),
-                value: base[i]
+        if (useRemove) {
+            inversePatches.unshift({
```

Comment:
Why are you using `unshift` here? And why is the inverse patch conditional when its associated patch is not?

Reply:
Remove operations shifts indices, so we add remove operations in reverse order. Inverse patch ignores when we can use replace 'length' property instead of multiple remove operations

Reply:
No. Add operations should be applied in ascend order, remove operations in descend

Reply:
Right, you would do it like this:
```js
for (let i = delta - 1; i >= 0; --i) {
  const path = basePath.concat([start + i])
  patches[i] = {
    op: "add",
    path,
    value: copy[start + i]
  }
  if (useRemove) {
    inversePatches.push({
      op: "remove",
      path
    })
  }
}
```